Tests/TuistCoreTests/Graph/GraphTraverserTests.swift

```diff
@@ -4601,6 +4601,67 @@ final class GraphTraverserTests: TuistUnitTestCase {
         XCTAssertEqual(result.1, platformCondition)
     }
 
+    // https://github.com/tuist/tuist/issues/5746
+    func test_transitiveTargetDependenciesWhenIntermediateDependenciesHaveConditions() throws {
```

Comment:
Thanks for including this 👍 

A few small suggestions:
- is it worth replicating the use case @alexanderwe reported (e.g. adding an extension in the example)? The scenario does replicate it but thought may be more clear to show a real use cases where you would specify a platform condition (e.g. from first glance I was thinking was the condition needed here :D) 
- For this test to catch the issue it needs to be run repeatedly, while possible in Xcode UI would be good to enforce it from the test _(we do something similar in the [stable generation test](https://github.com/tuist/tuist/blob/main/Tests/TuistIntegrationTests/Generator/StableStructureIntegrationTests.swift#L15))_


e.g.


```swift
// ...
// When
        let appToFrameworkCResults = (0..<100).map { _ in
            subject.combinedCondition(
                to: frameworkCDependency,
                from: appDependency
            )
        }

        let appToFrameworkDResults = (0..<100).map { _ in
            subject.combinedCondition(
                to: frameworkDDependency,
                from: appDependency
            )
        }

        // Then
        XCTAssertTrue(appToFrameworkCResults.allSatisfy { $0 == .condition(nil) })
        XCTAssertTrue(appToFrameworkDResults.allSatisfy { $0 == .condition(nil) })
```

Reply:
the product type of a target isn't taken into account at the graph traverser level.  if dependencies relationships are defined in the graph (which may take product types into account externally). This case could be replicated with frameworks that have capabilities that have platform specific features that rely on a system frameworks.   e.g. `FrameworkA` provides something for `macOS` relying on `StoreKit`, and then separately `FrameworkB` provides something on `iOS`  that relies on `StoreKit`.  if you have a Multiplatform (iOS/macOS) app that uses both frameworks, then `StoreKit` should be included in call cases without filters.  I don't think that the extension case needs to be part of this unit test.  Are you thinking any other places that could include something like this to help folks more generally understand conditions? 

for the second bit, I can add doing this a few times since it should be basically free due to caching now.

Reply:
to explain a little more since I'm sure this PR will be referenced in the future:

The bug manifested because `visited` would in some cases contain `FrameworkB` when computing the filters needed for `StoreKit`.  This would mean that only the filters of `FrameworkA` would be included in the union calculation and that would result in the wrong final condition being calculated.